File: python/16-3Sum Closest.py
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        "2 pointer, Time: O(n^2), Space: O(1)"
        n = len(nums)
        nums.sort()
        min_diff = float("+inf")
        ans = 0
        for i in range(n-2):
            l, r = i+1, n-1
            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if s == target: return s
                if abs(s-target) < min_diff:
                    min_diff = abs(s-target)
                    ans = s
                elif s < target: l += 1
                else: r -= 1
        return ans

nums = [-1,2,1,-4]
target = 1
ans = Solution().threeSumClosest(nums, target)
assert ans == 2

nums = [0,1,2]
target = 3
ans = Solution().threeSumClosest(nums, target)
assert ans == 3

File: cpp/16. 3Sum Closest.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
// Two Pointers, Time: O(n^2), Space: O(1)
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        int diff = INT_MAX;
        int ans;
        for(int i = 0; i < n-2; i++) {
            int l = i + 1, r = n-1;
            while(l < r) {
                int s = nums[i] + nums[l] + nums[r];
                if(s == target) return s;
                if(abs(s-target) < diff) {
                    diff = abs(s-target);
                    ans = s;
                } else if(s < target) {
                    l++;
                } else if(s > target) {
                    r--;
                }
            }
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution;
    vi nums = {-1,2,1,-4}; int target = 1;
    int ans = solution.threeSumClosest(nums, target);
    assert(ans==2);
}